rfcs/20210219-cache-compiled-gpu-kernels/README.md

```diff
@@ -0,0 +1,101 @@
+which has the 2nd state.
+
+There are two cons of such approach:
+1. Binary representation is basically an array of bytes hence no 3rd party resources
+are stored in the primitive cache
+2. Binary representation is associated with a particular GPU device but is context
+independent therefore it can be reused for different GPU contexts
+
+Current approach allows to avoid storing OpenCL and SYCL resources globally to
+avoid a potential issue which may result in a crash due to the order of unloading
+oneDNN and the libraries whose resource oneDNN holds globally.
+
+The main requirement of the approach is that only context independent resources
+can be stored in the primitive. To create and carry context dependent resources
+oneDNN has a resource abstraction which is created at each primitive creation stage
+regardless of whether it's a cache hit or cache miss.
+
+This approach has a number of drawbacks which become critical:
+1. There is overhead on creating a kernel out of a binary per each primitive creation
+    * That time can vary significantly depending on architecture. For example creating
+    kernels from binary for batch normalization on Gen9 takes ~200us but on a new
+    architecture it may take ~1.5ms. The reason is likely not optimized software
+    stack for that architecture but it means that such an issue can occur for each
+    new architecture
+2. Unfortunately memory is context dependent and hence cannot be stored in the
+primitive. The scales must be copied to the memory and the memory then
+is stored in the oneDNN resource. The copy also involves map/unmap operations
+3. SYCL is trying to abstract away backend specific behavior and if it cannot do that
+it gets rid of some backend specific API. For example not all SYCL abstractions allow
+to get native backend handle which can be used to work with backend API directly.
+In general, this problem is solvable. There is an option to request Intel extensions
+which will provide required API
+
+For the models that have small execution time, for example DLRM or BERT the
+overhead becomes critical. All optimizations that can be done at oneDNN side is
+completed. They allowed to get rid of overhead for models like ResNet-50 (MB 1024),
+but they didn't solve the problem.
+
+The only viable option is to get rid of the overhead completely. The proposal is
+to switch to caching compiled kernel rather than their binary representations.
+
+## How to deal with the unloading libraries issue?
+
+In general, it's seem unlikely to run into the issue because:
+1. According to the SYCL integration design, frameworks store GPU device and context
+abstractions globally. If that was an issue they wouldn't have done that
+2. Benchdnn and Gtests have global engine which contains GPU device and context
+abstractions and no issues so far
+
+If it becomes an issue, the workaround will be clearing the primitive cache by
+calling  `dnnl::set_primitive_cache_capacity(0);` before exit on the frameworks' side.
```

Comment:
This implies that the user knows implementation details (e.g. that the cache contains runtime objects).
Here I guess the proper workaround would be to introduce a `finalize` function. 

Also, please put that after the SYCL priority handling paragraph  (better optics) :-)

Reply:
> This implies that the user knows implementation details

Not necessarily, the only thing that user may need to know is that they may want to clear the cache before exiting. We can put a warning or a note in the documents.

I don't really like the finalize option because it's too general while the problem is currently limited to primitive cache. Having `finalize` function looks like a part of programming model where you need to finalize something without even initializing something. This looks counterintuitive to me.

There is already an API which could address that, why would we need to introduce a new one? Do you have in mind other situations where we may need to have finalize?